adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala

```diff
@@ -166,74 +194,175 @@ class Transform(protected val args: TransformArgs) extends BDGSparkCommand[Trans
+  /**
+   * @param rdd The RDD of reads that was output by [[maybeCoalesce]].
+   * @param sl The requested storage level for caching, if requested.
+   * @param sd A sequence dictionary describing the reference genome build.
+   * @return If -sortReads was set, returns the sorted reads. If
+   *   -sort_lexicographically is additionally set, sorts lexicographically,
+   *   instead of by contig index. If no sorting was requested, returns
+   *   the input RDD.
+   */
+  private def maybeSort(rdd: RDD[AlignmentRecord],
+                        sl: StorageLevel,
+                        sd: SequenceDictionary): RDD[AlignmentRecord] = {
     if (args.sortReads) {
-      val oldRdd = if (args.cache) {
-        adamRecords.persist(sl)
-      } else {
-        adamRecords
+
+      // cache the input if requested. sort is two stages:
+      // 1. sample to create partitioner
+      // 2. repartition and sort within partitions
+      if (args.cache) {
+        rdd.persist(sl)
       }
 
       log.info("Sorting reads")
-      if (args.sortLexicographically) {
-        adamRecords = oldRdd.sortReadsByReferencePosition()
+
+      // are we sorting lexicographically or using legacy SAM sort order?
+      val sortedRdd = if (args.sortLexicographically) {
+        rdd.sortReadsByReferencePosition()
       } else {
-        adamRecords = oldRdd.sortReadsByReferencePositionAndIndex(sd)
+        rdd.sortReadsByReferencePositionAndIndex(sd)
       }
 
+      // unpersist the cached rdd, if caching was requested
       if (args.cache) {
-        oldRdd.unpersist()
+        rdd.unpersist()
       }
+
+      sortedRdd
+    } else {
+      rdd
     }
+  }
 
+  /**
+   * @param rdd The RDD of reads that was output by [[maybeSort]].
+   * @param stringencyOpt An optional validation stringency.
+   * @return If -add_md_tags is set, recomputes the MD tags for the reads
+   *   provided. If some reads have tags, -md_tag_overwrite controls whether
+   *   these tags are recomputed or not. If MD tagging isn't requested, we
+   *   return the input RDD.
+   */
+  def maybeMdTag(rdd: RDD[AlignmentRecord],
+                 stringencyOpt: Option[ValidationStringency]): RDD[AlignmentRecord] = {
     if (args.mdTagsReferenceFile != null) {
       log.info(s"Adding MDTags to reads based on reference file ${args.mdTagsReferenceFile}")
-      adamRecords =
-        MDTagging(
-          adamRecords,
-          args.mdTagsReferenceFile,
-          fragmentLength = args.mdTagsFragmentSize,
-          overwriteExistingTags = args.mdTagsOverwrite,
-          validationStringency = stringencyOpt.getOrElse(ValidationStringency.STRICT)
-        )
+      MDTagging(
+        rdd,
+        args.mdTagsReferenceFile,
+        fragmentLength = args.mdTagsFragmentSize,
+        overwriteExistingTags = args.mdTagsOverwrite,
+        validationStringency = stringencyOpt.getOrElse(ValidationStringency.STRICT)
+      )
+    } else {
+      rdd
     }
+  }
+
+  def apply(rdd: RDD[AlignmentRecord],
```

Comment:
Agreed. I was quite pleased at the end. I was inspired by @ryan-williams's `maybeX` pattern, which we've got elsewhere.